### PR TITLE
Bernard/libvirt boilerplate

### PR DESCRIPTION
**Pull request by bernardjkim**
- Added libvirt boilerplate
- Copied over libvirt Terraform config from gravitational/gravity/terraform-dev.

### CODE REVIEW

File: assets/terraform/libvirt/config.tf [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+variable "image_name" {
+  description = "OS image file"
+  type        = "string"
+  default     = "ubuntu-16.04-server-cloudimg-amd64-disk1.img"
+}
+
+variable "gravity_dir_size" {
+  description = "Disk size for gravity directory"
```
Comment (r0mant) [inline assets/terraform/libvirt/config.tf]: `... in bytes`.

File: assets/terraform/libvirt/config.tf [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  type        = "string"
+  default     = "12000000000"
```
Comment (r0mant) [inline assets/terraform/libvirt/config.tf]: I think 12GB is not quite enough for Gravity data. I'd do at least 25-30.

File: infra/providers/libvirt/config.go [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package libvirt
+
+// Config specifies libvirt specific parameters
+type Config struct {
+	// SSHKeyPath specifies the location of the SSH private key for remote access
+	SSHKeyPath string `json:"-" yaml:"ssh_key_path" validate:"required"`
```
Comment (r0mant) [inline infra/providers/libvirt/config.go]: Why is this field ignored during json (un-)marshaling but not during yaml?

File: assets/terraform/libvirt/node.tf [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    - 'echo "/dev/vdc1 /tmp ext4 discard,noatime,nodiratime 0 0" >> /etc/fstab'
+    - 'mkdir -p /var/lib/gravity'
+    - 'mount -a'
+    - 'chmod 777 /tmp'
+    EOF
+}
+
+# Create the machine
+resource "libvirt_domain" "domain-gravity" {
+  name      = "telekube${count.index}"
```
Comment (r0mant) [inline assets/terraform/libvirt/node.tf]: The "telekube" name is obsolete -> "gravity".

File: assets/terraform/libvirt/node.tf [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+  memory    = "${var.memory_size}"
+  vcpu      = "${var.cpu_count}"
+  count     = "${var.nodes}"
+  cloudinit = "${libvirt_cloudinit_disk.commoninit.id}"
+
+  network_interface {
+    hostname        = "telekube${count.index}"
```
Comment (r0mant) [inline assets/terraform/libvirt/node.tf]: telekube -> gravity

File: assets/terraform/libvirt/output.tf [hunk 6]
```diff
@@ -0,0 +1,11 @@
+#
+# Output Variables
+#
+
+output "private_ips" {
+  value = "${libvirt_domain.domain-gravity.*.network_interface.0.addresses.0}"
+}
+
+output "public_ips" {
+  value = "${libvirt_domain.domain-gravity.*.network_interface.0.addresses.0}"
```
Comment (r0mant) [inline assets/terraform/libvirt/output.tf]: This seems like it will output exactly the same as "private_ips" above.

Comment (bernardjkim) [reply assets/terraform/libvirt/output.tf]: Yeah, I'm still not entirely sure what the public IPs are used for, but they are a required output.
Any advice on where to look to get a better understanding of how the public IPs are used in gravity?

File: infra/providers/libvirt/config.go (hunk 3 above)
Comment (bernardjkim) [reply infra/providers/libvirt/config.go]: I'm not quite sure. I copied over the field from the gce config. It looks like the gce and azure config both have SSHKeyPath ignored, but the aws config did not ignore it.
My guess is that this was meant as a security measure by not logging the location of the private SSH key?

Comment (r0mant) [reply infra/providers/libvirt/config.go]: Yeah, I was mostly wondering about why it is dumped during yaml marshaling :) Seems weird that if you do `json.Marshal` on this struct then this field will be dumped, but if you do `yaml.Marshal` it won't.

Comment (bernardjkim) [reply infra/providers/libvirt/config.go]: Oh I see. Yeah there are quite a few inconsistencies scattered throughout the code. Hopefully @a-palchikov can comment on this once hes settled in and had a chance to review.

File: assets/terraform/libvirt/output.tf (hunk 6 above)
Comment (a-palchikov) [reply assets/terraform/libvirt/output.tf]: Public IPs have been historically used to connect to nodes via SSH to run remote commands.

File: infra/gravity/config.go [hunk 7]
```diff
@@ -69,13 +70,15 @@ func (drv StorageDriver) Driver() string {
...
-	CloudProvider string `yaml:"cloud" validate:"required,eq=aws|eq=azure|eq=gce|eq=ops"`
+	CloudProvider string `yaml:"cloud" validate:"required,eq=aws|eq=azure|eq=gce|eq=libvirt|eq=ops"`
 	// AWS defines AWS connection parameters
 	AWS *aws.Config `yaml:"aws"`
 	// Azure defines Azure connection parameters
 	Azure *azure.Config `yaml:"azure"`
 	// GCE defines Google Compute Engine connection parameters
 	GCE *gce.Config `yaml:"gce"`
+	// LibVirt defines libvirt connection parameters
+	LibVirt *libvirt.Config `yaml:"libvirt"`
```
Comment (a-palchikov) [inline infra/gravity/config.go]: nit: just `Libvirt` would suffice.

File: infra/gravity/provision.go [hunk 8]
```diff
@@ -19,14 +19,14 @@ import (
 	sshutil "github.com/gravitational/robotest/lib/ssh"
 	"github.com/gravitational/robotest/lib/utils"
 	"github.com/gravitational/robotest/lib/wait"
+	uuid "github.com/satori/go.uuid"
```
Comment (a-palchikov) [inline infra/gravity/provision.go]: nit: I think it was in the correct group before.

File: infra/providers/libvirt/config.go (hunk 3 above)
Comment (a-palchikov) [reply infra/providers/libvirt/config.go]: Yes, JSON serialization is used for relaying the relevant attributes to `terraform apply` which the key path is not.

**Pull request merged**